File: Baekjoon Online Judge/hard/11438.java
import java.io.*;
import java.util.*;

public class Main {
    static List<List<Integer>> tree;
    static int[] depth;
    static int[][] parent;
    static int maxDepth;

    private static void dfs(int node, int parentNode) {
        depth[node] = depth[parentNode] + 1;
        parent[node][0] = parentNode;
        for (int i = 1; i <= maxDepth; ++i) {
            parent[node][i] = parent[parent[node][i - 1]][i - 1];
        }
        for (int child : tree.get(node)) {
            if (child != parentNode) {
                dfs(child, node);
            }
        }
    }

    private static int LCA(int u, int v) {
        if (depth[u] > depth[v]) {
            int temp = u;
            u = v;
            v = temp;
        }
        for (int i = maxDepth; i >= 0; --i) {
            if (depth[parent[v][i]] >= depth[u]) {
                v = parent[v][i];
            }
        }
        if (u == v) return u;

        int res = u;
        for (int i = maxDepth; i >= 0; --i) {
            if (parent[u][i] != parent[v][i]) {
                u = parent[u][i];
                v = parent[v][i];
            }
            res = parent[u][i];
        }
        return res;
    }

    public static void main(String[] args) throws IOException {
        // Init
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();

        // Input
        int N = Integer.parseInt(br.readLine());
        tree = new ArrayList<>();
        for (int i = 0; i <= N; ++i) {
            tree.add(new ArrayList<>());
        }
        for (int i = 0; i < N - 1; ++i) {
            st = new StringTokenizer(br.readLine());
            final int u = Integer.parseInt(st.nextToken());
            final int v = Integer.parseInt(st.nextToken());
            tree.get(u).add(v);
            tree.get(v).add(u);
        }

        maxDepth = (int) Math.ceil(Math.log(N) / Math.log(2));
        parent = new int[N + 1][maxDepth + 1];
        depth = new int[N + 1];
        dfs(1, 0);

        int M = Integer.parseInt(br.readLine());
        while (M-- > 0) {
            st = new StringTokenizer(br.readLine());
            final int u = Integer.parseInt(st.nextToken());
            final int v = Integer.parseInt(st.nextToken());

            // Solve
            sb.append(LCA(u, v)).append('\n');
        }

        // Output
        System.out.println(sb);
    }
}

File: Baekjoon Online Judge/hard/11438.py
from math import *
from heapq import heappop, heappush
from itertools import combinations
from collections import defaultdict, deque
from bisect import bisect_left, bisect_right
from sys import stdin, setrecursionlimit
from typing import List, Any

input = stdin.readline
setrecursionlimit(10 ** 6)

INF = int(1e10)


def dfs(cur, _parent):
    depth[cur] = depth[_parent] + 1

    # 희소배열 전처리
    parent[cur][0] = _parent
    for k in range(1, MAX_DEPTH + 1):
        parent[cur][k] = parent[parent[cur][k - 1]][k - 1]
    for child in tree[cur]:
        if child != _parent:
            dfs(child, cur)

def lca(u, v):
    if depth[u] < depth[v]:
        u, v = v, u
    for k in range(MAX_DEPTH, -1, -1):
        if depth[parent[u][k]] >= depth[v]:
            u = parent[u][k]
    if u == v:
        return u
    for k in range(MAX_DEPTH, -1, -1):
        if parent[u][k] != parent[v][k]:
            u = parent[u][k]
            v = parent[v][k]
    return parent[u][0]


if __name__ == '__main__':
    # Init
    tree = defaultdict(list)

    N = int(input())
    # Input
    for _ in range(N - 1):
        u, v = map(int, input().split())
        tree[u].append(v)
        tree[v].append(u)
    MAX_DEPTH = int(log2(N))
    parent = [[0] * (MAX_DEPTH + 1) for _ in range(N + 1)]
    depth = [0] * (N + 1)

    # Solve
    dfs(1, 0)

    res = []
    for _ in range(int(input())):
        u, v = map(int, input().split())
        res.append(lca(u, v))

    # Output
    print(*res, sep='\n')